userland/test-semaphore/main.c

```diff
@@ -0,0 +1,71 @@
+#include "userland.h"
+
+#define DEBUG(str) debug(str, sizeof(str))
+#define NUM_SEMS 3
+#define ASSERT(x) if (!(x)) { DEBUG("FAILED: FALSE: " # x); }
+#define ASSERT_OK(x) if (x) { DEBUG("FAILED: NONZERO: " # x); }
+
+semaphore_id sems[NUM_SEMS];
+void waker() {
+  nanosleep(500000000);
+  DEBUG("WAKE 0");
+  ASSERT_OK(sem_set(sems[0]));
+  yield();
+
+  /* sleeper is about to wait on 0 and 1, wake it twice */
+  DEBUG("WAKE 0 1");
+  ASSERT_OK(sem_set(sems[0]));
+  ASSERT_OK(sem_set(sems[1]));
+  yield();
+
+  /* sleeper is about to watch 2, set it */
+  DEBUG("WAKE 2");
+  ASSERT_OK(sem_set(sems[2]));
+}
+
+void sleeper() {
+  ASSERT_OK(sem_watch(sems[0]));
+  DEBUG("SLEEP 0");
+  /* test simple sleep-wake */
+  ASSERT(sem_wait() == sems[0]);
+  DEBUG("AWAKE 0");
+  ASSERT_OK(sem_watch(sems[1]));
+  yield();
+
+  /* test waking both of multiple watched sems */
+  DEBUG("SLEEP 0 1");
+  ASSERT(sem_wait() == sems[0]);
+  ASSERT(sem_wait() == sems[1]);
+  DEBUG("AWAKE 0 1");
+  yield();
+
+  /* test watching a semaphore that was already set */
+  DEBUG("WATCH 2");
+  ASSERT_OK(sem_watch(sems[2]));
+  DEBUG("SLEEP 0 1 2");
+  ASSERT(sem_wait() == sems[2]);
+  DEBUG("AWAKE 2");
+
+  for (int i = 0; i < NUM_SEMS; ++i) {
+    if (sem_delete(sems[i])) {
+      DEBUG("DELETION FAILED");
+    }
+  }
+}
+
+void main() {
```

Comment:
The code path is not very interesting since we'll never be WAITING during the set. I think a better test would be unwatch but there's no way to definitively test it without some kind of timeout or something.